perf/src/sigverify.rs

```diff
@@ -420,34 +423,78 @@ pub fn generate_offsets(
     )
 }
 
-fn dedup_packet(count: &AtomicU64, packet: &mut Packet, bloom: &AtomicBloom<&[u8]>) {
-    // If this packet was already marked as discard, drop it
-    if packet.meta.discard() {
-        return;
+pub struct Deduper {
+    filter: Vec<AtomicU64>,
+    seed: RandomState,
+    age: Instant,
+    max_age: Duration,
+    saturated: AtomicBool,
+}
+
+impl Deduper {
+    pub fn new(size: u32, max_age_ms: u64) -> Self {
+        let mut filter: Vec<AtomicU64> = Vec::with_capacity(size as usize);
+        filter.resize_with(size as usize, Default::default);
+        let seed = RandomState::new();
+        Self {
+            filter,
+            seed,
+            age: Instant::now(),
+            max_age: Duration::from_millis(max_age_ms),
+            saturated: AtomicBool::new(false),
+        }
     }
 
-    // If this packet was not newly added, it's a dup and should be discarded
-    if !bloom.add(&&packet.data.as_slice()[0..packet.meta.size]) {
-        packet.meta.set_discard(true);
-        count.fetch_add(1, Ordering::Relaxed);
+    pub fn reset(&mut self) {
+        let now = Instant::now();
+        let saturated = self.saturated.load(Ordering::Relaxed);
+        if saturated || now.duration_since(self.age) > self.max_age {
+            for i in &self.filter {
+                i.store(0, Ordering::Relaxed);
+            }
+            self.seed = RandomState::new();
+            self.age = now;
+            self.saturated.store(false, Ordering::Relaxed);
+        }
     }
-}
 
-pub fn dedup_packets(bloom: &AtomicBloom<&[u8]>, batches: &mut [PacketBatch]) -> u64 {
-    use rayon::prelude::*;
-    let packet_count = count_packets_in_batches(batches);
-    // machine specific random offset to read the u64 from the packet signature
-    let count = AtomicU64::new(0);
-    PAR_THREAD_POOL.install(|| {
-        batches.into_par_iter().for_each(|batch| {
-            batch
-                .packets
-                .par_iter_mut()
-                .for_each(|p| dedup_packet(&count, p, bloom))
-        })
-    });
-    inc_new_counter_debug!("dedup_packets_total", packet_count);
-    count.load(Ordering::Relaxed)
+    fn dedup_packet(&self, count: &AtomicU64, packet: &mut Packet) {
+        // If this packet was already marked as discard, drop it
+        if packet.meta.discard() {
+            return;
+        }
+        let mut hasher = self.seed.build_hasher();
+        hasher.write(&packet.data[0..packet.meta.size]);
+        let hash = hasher.finish();
+        let len = self.filter.len();
+        let pos = (usize::try_from(hash).unwrap()).wrapping_rem(len);
+        // saturate each position with or
+        let prev = self.filter[pos].fetch_or(hash, Ordering::Relaxed);
+        if prev == u64::MAX {
```

Comment:
Presumably we get here pretty quickly?

Reply:
not really, on a 1m unit haven't seen it hit in 1m tries of random values.  peak hill is going to take a while to hit with random bits.  false positive rate goes up to 1/1000 in 1m packets on 1m sized array

Reply:
Clearly I have no intuition for these things :sweat_smile: 

Would it make sense to reset the entry to the current hash instead of zero?